packages/react-reconciler/src/__tests__/ReactIncrementalErrorReplay-test.js

```diff
@@ -0,0 +1,59 @@
+/**
+ * Copyright (c) 2013-present, Facebook, Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @emails react-core
+ * @jest-environment node
+ */
+
+'use strict';
+
+let React;
+let ReactNoop;
+
+describe('ReactIncrementalErrorReplay', () => {
+  beforeEach(() => {
+    jest.resetModules();
+    React = require('react');
+    ReactNoop = require('react-noop-renderer');
+  });
+
+  function div(...children) {
+    children = children.map(c => (typeof c === 'string' ? {text: c} : c));
+    return {type: 'div', children, prop: undefined};
+  }
+
+  function span(prop) {
+    return {type: 'span', children: [], prop};
+  }
+
+  it('should fail gracefully on error in the host environment', () => {
+    ReactNoop.simulateErrorInHostConfig(() => {
+      ReactNoop.render(<span />);
+      expect(() => ReactNoop.flush()).toThrow('Error in host config.');
+    });
+  });
+
+  it('should fail gracefully on error that does not reproduce on replay', () => {
+    let didInit = false;
+
+    function badLazyInit() {
+      const needsInit = !didInit;
+      didInit = true;
+      if (needsInit) {
+        throw new Error('Hi');
+      }
+    }
+
+    class App extends React.Component {
+      render() {
+        badLazyInit();
+        return <div />;
+      }
+    }
+    ReactNoop.render(<App />);
+    expect(() => ReactNoop.flush()).toThrow('Hi');
```

Comment:
This one fails because we assume rendering is idempotent:

https://github.com/facebook/react/blob/6b99c6f9d376bacbb769264d743c405b495b03ad/packages/react-reconciler/src/ReactFiberScheduler.js#L297-L298

But there *are* cases (like lazy init) where it might throw the first time but not on subsequent calls. We should handle this gracefully. When this happens, when we exit this block, `failedUnitOfWork` is not the same as `nextUnitOfWork` (which might be null):

https://github.com/facebook/react/blob/6b99c6f9d376bacbb769264d743c405b495b03ad/packages/react-reconciler/src/ReactFiberScheduler.js#L939